component/kafka/simple/consumer.go

```diff
@@ -0,0 +1,300 @@
+	}
+}
+
+func (c *consumerHandler) consume(ctx context.Context) (hasProcessedMessages bool, err error) {
+	eg, ctx := errgroup.WithContext(ctx)
+	for partition, pc := range c.pcs {
+		pc := pc
+		partition := partition
+		eg.Go(func() error {
+			return c.consumePartition(ctx, pc.Messages(), partition)
+		})
+	}
+	return c.hasProcessedMessages, eg.Wait()
+}
+
+func (c *consumerHandler) consumePartition(ctx context.Context, ch <-chan *sarama.ConsumerMessage, partition int32) error {
+	if c.component.notificationOnceReachingLatestOffset() {
+		// We don't want to wait for consuming a message if we already know we're at the end of the stream
+		if c.startingOffsetAfterLatestOffset(c.component.latestOffsets[partition], partition) {
+			c.mu.Lock()
+			c.partitionHasReachedLatestOffset(partition)
+			c.mu.Unlock()
+		}
+	}
+
+	if c.isBatch {
+		return c.consumePartitionBatch(ctx, ch, partition)
+	}
+
+	return c.consumePartitionUnit(ctx, ch, partition)
+}
+
+func (c *consumerHandler) startingOffsetAfterLatestOffset(latestOffset int64, partition int32) bool {
+	return c.component.startingOffsets[partition] >= latestOffset
+}
+
+func (c *consumerHandler) partitionHasReachedLatestOffset(partition int32) {
+	delete(c.partitionsWithLatestOffsetUnreached, partition)
+	if len(c.partitionsWithLatestOffsetUnreached) == 0 {
+		c.notificationAlreadySent = true
+		// As the consumer can be retried, we have to make sure the channel is closed only once.
+		c.component.once.Do(func() {
+			close(c.component.latestOffsetReachedChan)
+		})
+	}
+}
+
+func (c *consumerHandler) consumePartitionUnit(ctx context.Context, ch <-chan *sarama.ConsumerMessage, partition int32) error {
+	for {
+		select {
+		case msg, ok := <-ch:
+			if !ok {
+				log.Debug("messages channel closed")
+				return nil
+			}
+			log.Debugf("message claimed: value = %s, timestamp = %v, partition = %d topic = %s", string(msg.Value), msg.Timestamp, msg.Partition, msg.Topic)
+			messageStatusCountInc(messageReceived, partition, msg.Topic)
+			err := c.unit(ctx, msg)
+			if err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			if ctx.Err() != context.Canceled {
+				log.Infof("closing consumer: %v", ctx)
+			}
+			return nil
+		}
+	}
+}
+
+func (c *consumerHandler) consumePartitionBatch(ctx context.Context, ch <-chan *sarama.ConsumerMessage, partition int32) (err error) {
+	// We need to enforce flushing the buffer before to return (e.g., context cancelled but messages are still buffered)
+	defer func() {
+		errFlush := c.flush(ctx)
+		if errFlush != nil {
+			err = errFlush
+		}
+	}()
+
+	for {
+		select {
+		case msg, ok := <-ch:
+			if ok {
+				log.Debugf("message claimed: value = %s, timestamp = %v, partition = %d topic = %s", string(msg.Value), msg.Timestamp, msg.Partition, msg.Topic)
+				messageStatusCountInc(messageReceived, partition, msg.Topic)
+				err := c.insertMessage(ctx, msg)
+				if err != nil {
+					return err
+				}
+			} else {
+				log.Debug("messages channel closed")
```

Comment:
Can you please elaborate on when this happens? e.g. this is a normal condition (I suppose)
Perhaps we could add some comment as well?